cloudmarker/clouds/gcpcloud.py

```diff
@@ -71,213 +88,186 @@ def _get_resource(self, service_name, version='v1'):
 
-    def _get_firewalls(self, project_id):
-        """Query the compute resource and return firewall rules.
+    def _get_projects(self):
+        """Generate tuples of record types and projects.
 
-        Arguments:
-            project_id (str): ID of the project for which firewall rules
-                have to be obtained.
+        The yielded tuples when unpacked would become arguments for
+        :meth:`_get_resources`. Each such tuple represents a single unit
+        of work that :meth:`_get_resources` can work on independently in
+        its own worker thread.
 
-        Returns:
-            list: A list of :obj:`dict` objects where each :obj:`dict`
-                object contains a firewall rule.
+        Yields:
+            tuple: A tuple which when unpacked forms valid arguments for
+                :meth:`_get_resources`.
 
         """
-        # Get firewall resource from the compute resource
-        firewall_resource = self._compute_resource.firewalls()
+        try:
+            cloud_resource = self._build_resource('cloudresourcemanager', 'v1')
+            compute_resource = self._build_resource('compute', 'v1')
 
-        firewall_rules = []
-        next_page_token = None
+            projects = _get_resource_iterator(cloud_resource.projects(),
+                                              'projects')
 
-        while True:
-            # Prepare the request to get the list of all firewall rules for a
-            # project and execute that request.
-            firewall_rules_object = firewall_resource.list(
-                project=project_id,
-                pageToken=next_page_token).execute()
+            for project_index, project in enumerate(projects):
+                project_id = project.get('projectId')
 
-            # firewall_rules_object contains firewall rules for the project
-            # if it has the 'items' key.
-            firewall_rules.extend(firewall_rules_object.get('items', []))
+                _log.info('Found %s',
+                          util.outline_gcp_project(project_index, project,
+                                                   None, self._key_file_path))
 
-            # firewall_rules_object will only contain the maximum number of
-            # results per page which is specified in options keyword argument
-            # maxResults in list() API call. The default value of maxResults is
-            # 500, if the result contains more than 500 records, nextPageToken
-            # is present in response whose value is to be used to make the
-            # subsequent list() call.
-            if 'nextPageToken' not in firewall_rules_object.keys():
-                break
+                yield ('firewall', project_index, project)
 
-            next_page_token = firewall_rules_object['nextPageToken']
+                zones = _get_resource_iterator(compute_resource.zones(),
+                                               'items', project=project_id)
```

Comment:
While testing, I found a problem here. If the zone iteration fails for any reason, then the `for project_index, project in enumerate(projects)` loop crashes before it can pull the remaining projects. A fix for this should go into the `_get_resource_iterator()` function.